### ISSUE TITLE
Discretization and Categorical features

### ISSUE THREAD

**Issue by Dola47** [marcotcr/lime#485]
Hi,

I have some questions that I will be glad to have some precise answers to them:

1- What are the benefits of applying discretization in the case of LimeTabularExplainer, and what are the disadvantages? In other words, what I will gain if I applied the discretization and what I will lose if I have not applied it?

2- I tried to implement all of the discretizers in LIME, each one result in different explanations. Accordingly, I need to know what is the best discretizer out of them, and why the quartile discretizer has been chosen to be the default one?

3- Why the categorical features have to be LabelEncoded? 

>  Finally, we use a One-hot encoder, so that the classifier does not take our categorical features as continuous features. We will use this encoder only for the classifier, not for the explainer - and the reason is that the explainer must make sure that a categorical feature only has one value.

What does it mean that the explainer must make sure that a categorical feature only has one value? Does LIME sum back the different categories from the same categorical feature in case of OneHotEncoding, or it treats them as if each one is a  new separate feature?

Thanks.

**Comment by marcotcr** [comment]
1 and 2: See #196, #113, #265. Different discretizations give you different 'interpretable features', from which you get (different) explanations. You need a granularity that makes sense to the user, but also is able to capture model behavior. The straw man is the discretization that says 'every value falls into the same bin'. Clearly, we won't get useful explanations out of that. The other extreme is saying 'either the value is =5.3323 or it's not', which is also not very useful.

> What does it mean that the explainer must make sure that a categorical feature only has one value? Does LIME sum back the different categories from the same categorical feature in case of OneHotEncoding, or it treats them as if each one is a new separate feature?

If you have a variable that encodes Sex (Male or Female) as a categorical feature, you want to generate perturbed samples of people who are either male or female. If you use OneHotEncoding, you get a feature for male and one for female, and LIME has no way of knowing that they are tied such that only one of them can be on.

**Comment by Dola47** [comment]
@marcotcr I got what you say. However, I can not really interpret the results out of discretized LIME, In other words, what exactly each feature mean, can you describe the results in this image https://user-images.githubusercontent.com/45386621/78400902-d76f7800-75f7-11ea-9286-108e886df292.PNG

**Comment by marcotcr** [comment]
1. I don't know if I can explain it better than I did [here](https://github.com/marcotcr/lime/issues/113#issuecomment-337657003). Did you have any specific questions that are not addressed there?

2. We pick one of the bins according to its frequency in the training set. Then, we generate a value in that bin by a truncated norm within that bin. Thus, no, categorical features will not be normally distributed. You can think of the simple case where the training set is distributed in the following way, assuming quartile discretization: [0.1, 0.1, 0.1, 0.7]. When sampling from this feature, 70% of values will fall into the last quartile, 10% in each of the other quartiles. I guess it will be a mixture of truncated normals.

3 - I don't think I understand what you mean by ' in a lower feature space than the original space as well.' Sorry!

**Comment by Dola47** [comment]
@marcotcr  For point three, what I mean by lower feature space, for example, if the original space work on 345 features, the interpretable space may work on 100 features. For example, for Images, we might have the original dimension of 100*100*100, and the lower interpretable space works only on 50*50*50 is the same holds for TabularData?

**Comment by marcotcr** [comment]
No, for tabular data we use the same features. Discretization changes the values that the features can take, but it's still the same features.

**Comment by Dola47** [comment]
@marcotcr  So to sum up, x' only represents a binary interpretable version but with the same number of features. What I see also that the similarity is done between x' and z' both as a binary interpretable version, not in between x and z in the original space, is that correct? 

I will be glad if you can have a look at the illustration of the algorithm I provided I think the only difference that x' should be considered only as a binary interpretable feature (not in lower feature space).

Moreover, can you have a look at point 5 regarding the ridge regression and inform me if my understanding is correct or not?

If we applied the discretization, does it will be applied to all of the features (numerical and categorical) or to the categorical only?

For your comment here, I think you by ([0.1, 0.1, 0.1, 0.7] that this is how only one feature is discredited. Accordingly, we gonna pick the last bin because it has the highest frequency and we will generate value from it by using a truncated norm, is that correct?
> 2. We pick one of the bins according to its frequency in the training set. Then, we generate a value in that bin by a truncated norm within that bin. Thus, no, categorical features will not be normally distributed. You can think of the simple case where the training set is distributed in the following way, assuming quartile discretization: [0.1, 0.1, 0.1, 0.7]. When sampling from this feature, 70% of values will fall into the last quartile, 10% in each of the other quartiles. I guess it will be a mixture of truncated normals.

**Comment by marcotcr** [comment]
5. In practice, what we do is first select K features (to make sure the constraint is met) with whatever feature selection method is picked (lasso, highest weights, or whatever), and then learn a ridge regressor with small regularization (for numerical stability) to get the actual weights for explanations.

> If we applied the discretization, does it will be applied to all of the features (numerical and categorical) or to the categorical only?  

Only to continuous features

> For your comment here, I think you by ([0.1, 0.1, 0.1, 0.7] that this is how only one categorical feature is discredited. Accordingly, we gonna pick the last bin because it has the highest frequency and we will generate value from it by using a truncated norm, is that correct?

Yes, this is for one feature (notice I'm not telling you what the bins are, just that they are quartiles). We are **not** going to pick the last bin, we will sample the bin according to these probabilities (i.e. the last bin will be picked ~70% of the time)

**Comment by Dola47** [comment]
> 1. In practice, what we do is first select K features (to make sure the constraint is met) with whatever feature selection method is picked (lasso, highest weights, or whatever), and then learn a ridge regressor with small regularization (for numerical stability) to get the actual weights for explanations.
Yup, but even after picking our K features, we might have several ridge regressors that fit and we will pick the one that contributes the least to the loss, so G here can be a set of several ridge aggressors, is that correct? Or is it only one ridge regressor that fits perfectly to our situation and what we mean by g belong to G, that is g can be considered for example as a linear model or decision tree, but it is only one model not several from the same type?

> Yes, this is for one feature (notice I'm not telling you what the bins are, just that they are quartiles). We are **not** going to pick the last bin, we will sample the bin according to these probabilities (i.e. the last bin will be picked ~70% of the time)

Ahh, so for this specific feature, what is going to be done, is that we gonna truncate from all of the bins and the final result for these features will be the result of multiplication of all of these truncated norms from each of the bins.

Something more, can you provide me how can we describe the interpretable version x' for the case of tabular data? In the case of Images, it might be superpixels, in case of text it will be words, how should it be for tabular data?

**Comment by marcotcr** [comment]
On Ridge: yes, we only have a single model that optimizes our locally weighted loss, out of the class of all ridge regressors. g belonging to G means that, yes.

For tabular data, you can think of the discretized version (if you use it) as the interpretable representation. If you don't use it, the original representation is already interpretable, and thus x = x'.

**Comment by Dola47** [comment]
> For tabular data, you can think of the discretized version (if you use it) as the interpretable representation. If you don't use it, the original representation is already interpretable, and thus x = x'.

So, if we applied the discretization, we can say that the similarity is done between x' and z' which are discretized versions. If the discretization is false then the similarity is done between x and z which are the same as the original version, correct?

**Comment by FrugoFruit90** [comment]
Please note that LabelEncoder is not meant to be used with categorical features. 
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html
I'm afraid this might not work well with Sklearn's Pipeline.
